### PR TITLE
Organize test jobs

### PR DESCRIPTION
**Pull request by sgugger**
# What does this PR do?

This PR reorganizes the jobs run on CircleCI to avoid launching setups and running the test fetcher multiple times when there is no tests found to run.

More precisely, it always run the three following jobs:
- `check_code_quality`
- `check_repo_consistency`
- `fetch_tests`
Then the actual tests jobs are run after `fetch_tests` and are immediately cancelled graciously if the `fetch_tests` job didn't find any tests. All of them read the list of tests found by `fetch_tests` and don't re-run the test_fetcher.

This also introduces a `fetch_all_tests` job which creates the file all jobs are looking for and fill it with all the tests, so the nightly run can reuse the same jobs as the standard one. As a result, all jobs `xxx_all` can be safely deleted.

To see the result of a run with one py modified, look at this [report](https://github.com/huggingface/transformers/runs/8381129029) (commit [With a modification in one file only](https://github.com/huggingface/transformers/pull/19058/commits/44c3a39d420a9118dc3586104fe0d3fcaf7c2321) below). Some of the tests are run only on the impacted tests like others (examples, custom tokenizers, layout lm tests) are run on specific tests as long as there was at least a modification warranting some tests. This is the same behavior as before.

To see the result of a run with no code modification, look at this [report](https://github.com/huggingface/transformers/pull/19058/checks?check_run_id=8381322343) (commit [No change, no tests](https://github.com/huggingface/transformers/pull/19058/commits/4497731b550ee7db9f82f0d6d5384e352b51b904) below). All jobs are still run but take a couple of seconds only.

To see the result of a run with all tests run, look at the CI report of this PR (which cleans up a modification I added in the setup and merged by mistake).

### CODE REVIEW

**Comment by LysandreJik** [comment]
I think your Build PR Documentation is hanging again :eyes:

File: .circleci/config.yml [hunk 1]
```diff
@@ -62,48 +62,50 @@ references:
...
+    # To run all tests for the nightly build
+    fetch_all_tests:
+        working_directory: ~/transformers
+        docker:
+            - image: cimg/python:3.7.12
+        parallelism: 1
+        steps:
+            - run: |
+                  mkdir test_preparation
+                  echo "tests" > test_preparation/test_list.txt
```
Comment (LysandreJik) [inline .circleci/config.yml:101]: Smart!

File: .circleci/config.yml [hunk 2]
```diff
@@ -116,6 +118,13 @@ jobs:
         parallelism: 1
         steps:
             - checkout
+            - attach_workspace:
+                at:  ~/transformers/test_preparation
+            - run: |
+                if [ ! -s test_preparation/test_list.txt ]; then
+                    echo "No tests to run, exiting early!"
+                    circleci-agent step halt
+                fi
```
Comment (LysandreJik) [inline .circleci/config.yml:127]: So given this, the workflow will always be there, but it will sometimes last ~0 seconds, I suppose?

**Review by LysandreJik — APPROVED** [review]
Much cleaner to not have everything twice. Great, this will save a lot of compute! Thanks for spending time working on it, @sgugger.

File: .circleci/config.yml [hunk 3]
```diff
@@ -1119,6 +730,22 @@ jobs:
                   perform-login: true
             - *delete_gke_jobs
 
+    run_tests:
```
Comment (ydshieh) [inline .circleci/config.yml:733]: I can't find when this job is run.

File: .circleci/config.yml [hunk 4]
```diff
@@ -116,6 +118,13 @@ jobs:
         parallelism: 1
         steps:
             - checkout
+            - attach_workspace:
+                at:  ~/transformers/test_preparation
```
Comment (ydshieh) [inline .circleci/config.yml:122]: Any reason for 2 spaces after `at:` ?

Comment (ydshieh) [reply .circleci/config.yml:122]: Same question for other jobs

**Review by ydshieh — APPROVED** [review]
You are the saver, @sgugger ! LGTM, thanks a lot.

File: .circleci/config.yml (hunk 2 above)
Comment (sgugger) [reply .circleci/config.yml:127]: 11s to be exact (it spins up and does the checkout, which can't be done after attaching the folder).

File: .circleci/config.yml (hunk 3 above)
Comment (sgugger) [reply .circleci/config.yml:733]: Ah ah, I was using it for my internal testing. Will remove!

File: .circleci/config.yml (hunk 4 above)
Comment (sgugger) [reply .circleci/config.yml:122]: Copy paste from their doc. It should work the same with just one space.

**Pull request merged**